### PR TITLE
m4rie feature and build system support

### PR DESCRIPTION
**Pull request by orlitzky**
Add a feature in `sage.features.m4rie` that has support from our build system. Minor changes to `MatrixSpace()` are needed, then  we add a few `# needs m4rie` tags, and test it all on the minimal CI job.

This one's easy.

### CODE REVIEW

**Comment by orlitzky** [comment]
@isuruf @saraedum does the linbox in conda really depend on m4rie (https://github.com/conda-forge/linbox-feedstock/blob/main/recipe/meta.yaml#L43)? I checked the tarball but did not find anything:

```
linbox-1.7.1 $ grep -ir m4ri
ChangeLog:	* M4RI checks
linbox/matrix/dense-matrix.h:// #include "linbox/matrix/densematrix/m4ri-matrix.h"
linbox/matrix/matrix-domain.h:// #include "matrixdomain/matrix-domain-m4ri.h"
linbox/matrix/matrixdomain/matrix-domain-gf2.h:	 * @todo this is where m4ri will play.
tests/test-matrix-domain.h:M4riDomain
```

**Comment by cxzhong** [comment]
@orlitzky can you resolve conflict?

**Comment by orlitzky** [comment]
> @orlitzky can you resolve conflict?

Done, thanks. It would be better to avoid the `--force` with `conda remove`... I'll open an issue on the conda repo.

File: src/sage/matrix/matrix_space.py
Comment (cxzhong) [inline src/sage/matrix/matrix_space.py]: This seems a failure

Comment (orlitzky) [reply src/sage/matrix/matrix_space.py]: Thanks, I have removed it (the same test appears further down).

Do you know how to update the conda env to pull in the latest linbox (not needing m4rie)? I would rather not merge this with the ugly CI hack.

Comment (cxzhong) [reply src/sage/matrix/matrix_space.py]: > Thanks, I have removed it (the same test appears further down).
> 
> Do you know how to update the conda env to pull in the latest linbox (not needing m4rie)? I would rather not merge this with the ugly CI hack.

you means the conda-lock files, you just run `python tools/update-conda.py`

**Comment by orlitzky** [comment]
On 2026-08-13 09:56:47, Chenxin Zhong wrote:
>
> you means the conda-lock files, you just run `python tools/update-conda.py`
>

I need a way to do it without installing conda because it doesn't support my CPU :\

**Comment by cxzhong** [comment]
> On 2026-08-13 09:56:47, Chenxin Zhong wrote: you means the conda-lock files, you just run `python tools/update-conda.py`
> I need a way to do it without installing conda because it doesn't support my CPU :\

Myabe you need to create a python venv and install grayskull via pip. Then you can run `update-conda.py` in this env

**Comment by orlitzky** [comment]
> Myabe you need to create a python venv and install grayskull via pip. Then you can run `update-conda.py` in this env

Doesn't work unfortunately, I get a `NotImplementedError` complaining about jinja2 (possibly missing plugin?). I'll just wait for somebody else to do it :P

**Comment by isuruf** [comment]
Here's a commit https://github.com/isuruf/sage/commit/006c73855743217acedd0f00ff46b59c8c4ebe0a

**Comment by orlitzky** [comment]
> Here's a commit [isuruf@006c738](https://github.com/isuruf/sage/commit/006c73855743217acedd0f00ff46b59c8c4ebe0a)

Many thanks :pray:

**Comment by cxzhong** [comment]
you can use https://github.com/sagemath/sage/pull/41879/commits/c2b1651f93e70550f5dd3d3b2ad527a2f9e92143
and https://github.com/sagemath/sage/pull/41879/commits/412ed5ac69c34cf8994589cf4ac55f60142b55c3 to fix type errors

**Comment by cxzhong** [comment]
The Meson failures are caused by the refreshed Conda lockfiles, not by the m4rie changes themselves.

The refresh changed the Linux compiler stack from `c-compiler` / `cxx-compiler` 1.11 together with the `gcc_linux-64` / `gxx_linux-64` activation packages to `c-compiler` / `cxx-compiler` 2.0 with the native `gcc` / `gxx` packages. With the new stack, `CC` and `CXX` are not populated during Conda activation.

However, `.github/workflows/ci-meson.yml` still unconditionally does:

```bash
export CC="ccache $CC"
export CXX="ccache $CXX"
```

so the resulting compiler commands are just `ccache`. Meson then correctly stops with:

```text
ERROR: Compiler cache specified without compiler: ccache
```

All of the Ubuntu matrix jobs fail at this configuration/metadata stage, before any Sage or m4rie source is compiled. The simultaneous Meson 1.12 update is not the root cause; older Meson versions also reject a cache-only compiler command.

A minimal forward-compatible fix would be:

```bash
if [[ -n "${CC:-}" ]]; then export CC="ccache $CC"; fi
if [[ -n "${CXX:-}" ]]; then export CXX="ccache $CXX"; fi
```

The ccache wrapper directories are already prepended to `PATH`, so when those variables are unset, Meson can auto-detect the compiler through the wrappers.

**Comment by orlitzky** [comment]
I appreciate the help, but I think I will just go back to the `--force` for now. That way the changes are limited to the minimal CI job (the thing I'm changing). I can always remove the `--force` later on, since I have to edit that line every time I add build system support for a new optional dependency.

File: src/sage/matrix/matrix_space.py [hunk 1]
```diff
@@ -2876,6 +2877,14 @@ def _MatrixSpace_ZZ_2x2():
...
-register_unpickle_override('sage.matrix.matrix_mod2e_dense',
-    'unpickle_matrix_mod2e_dense_v0', unpickle_matrix_gf2e_dense_v0)
+
+def _unpickle_gf2e(a, base_ring, nrows, ncols):
+    from sage.features.m4rie import M4rie
+    M4rie().require()
+    from sage.matrix.matrix_gf2e_dense import unpickle_matrix_gf2e_dense_v0
+    return unpickle_matrix_gf2e_dense_v0(a, base_ring, nrows, ncols)
+
+register_unpickle_override(
```
Comment (cxzhong) [inline src/sage/matrix/matrix_space.py:2887]: [P2] Guard current-format M4RIE pickles as well

This override protects only the legacy global `sage.matrix.matrix_mod2e_dense.unpickle_matrix_mod2e_dense_v0`, while `Matrix_gf2e_dense.__reduce__()` still emits `sage.matrix.matrix_gf2e_dense.unpickle_matrix_gf2e_dense_v0`. If m4rie is hidden, or disabled while an extension from a previous build remains importable (the case documented by the new feature), `loads()` therefore reaches the unguarded current function. It creates the now-default generic `MatrixSpace` and then attaches a `Matrix_gf2e_dense` element to it. I reproduced `type(B) is B.parent().Element == False`; `B + B.parent().zero()` then aborts in `mzed_add` with SIGABRT. Please enforce `M4rie().require()` and construct an explicit m4ri parent in the actual unpickler. Registering the current global here would protect Sage's unpickler, but fixing the actual function is also needed for stdlib pickle and direct reduce-function calls.

File: src/meson.build [hunk 2]
```diff
@@ -176,11 +176,11 @@ if not m4ri.found()
...
 endif
-m4rie = dependency('m4rie', required: false)
-if not m4rie.found()
-  # For some reason, m4rie is not found via pkg-config on some systems (eg Conda)
-  m4rie = cc.find_library('m4rie', required: not is_windows, disabler: true)
-endif
+
+# m4rie requires m4ri, but the check should fail if m4ri is not present,
+# so let's not over-complicate the logic.
+m4rie = dependency('m4rie', required: get_option('m4rie'), disabler: true)
```
Comment (cxzhong) [inline src/meson.build:182]: [P2] Preserve detection of supported m4rie installs without metadata

`dependency('m4rie')` can only use dependency metadata here, so this drops the previous library fallback. Sage's distro configure still accepts a system m4rie using only `m4rie/m4rie.h` plus `gf2e_init`, with no minimum version; for example, upstream release-20150908 satisfies that check but predates `m4rie.pc`. Such a build now declines to install the bundled SPKG and then Meson silently disables `matrix_gf2e_dense` in auto mode (or errors with `-Dm4rie=enabled`) despite the accepted library being usable. Please keep a feature-option-aware `cc.find_library()` fallback, or update the configure-side requirement so these installations are consistently rejected and replaced.

File: src/sage/features/m4rie.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    def __init__(self):
+        r"""
+        EXAMPLES::
+
+            sage: from sage.features.m4rie import M4rie
+            sage: M4rie()
+            Feature('m4rie')
+
+        """
+        module_name = "sage.matrix.matrix_gf2e_dense"
```
Comment (cxzhong) [inline src/sage/features/m4rie.py:43]: [P2] Split M4RIE out of the existing `sage.libs.m4ri` feature

Adding this independent feature leaves `sage.features.sagemath.sage__libs__m4ri` as a `JoinFeature` that still requires both `matrix_gf2e_dense` and `matrix_mod2_dense`. With `-Dm4rie=disabled`, `matrix_mod2_dense` is built and usable, but `sage.libs.m4ri` reports false. The minimal CI log shows 488 `matrix_mod2_dense.pyx` doctests running while `sage.libs.m4ri` is absent from the detected-feature list; this also skips four pure-M4RI examples in `matrix_space.py` and `matrix2.pyx`. Please make `sage.libs.m4ri` check only `matrix_mod2_dense`, update its description, and retag the genuinely M4RIE-dependent example in `matrix/special.py` as `needs m4rie`.

**Review by cxzhong — CHANGES_REQUESTED** [review]
Thanks for working on making m4rie independently configurable. I found three issues that should be fixed before merge: current-format M4RIE pickles can produce an invalid parent/element pairing and abort the process when the feature is unavailable; the Meson change drops support for system m4rie installations without dependency metadata that Sage's configure step still accepts; and the existing sage.libs.m4ri feature becomes a false negative when only M4RIE is disabled. I reproduced the pickle crash on this head (SIGABRT in mzed_add). The current CI is green, but the minimal job also demonstrates the sage.libs.m4ri false negative through its detected-feature list and skipped doctests.

File: src/meson.build (hunk 2 above)
Comment (orlitzky) [reply src/meson.build:182]: m4rie.pc was added in 2019: https://github.com/malb/m4rie/commit/d832744df1c5bf6afe506a2b7c1543c810752f5f

The issue with conda is fixed, and I don't think we need to support the 2015 release...

The sage distro is also using m4rie-20250128, and should install a working pc file?

File: src/sage/matrix/matrix_space.py (hunk 1 above)
Comment (orlitzky) [reply src/sage/matrix/matrix_space.py:2887]: > [P2] Guard current-format M4RIE pickles as well
> 
> This override protects only the legacy global `sage.matrix.matrix_mod2e_dense.unpickle_matrix_mod2e_dense_v0`, while `Matrix_gf2e_dense.__reduce__()` still emits `sage.matrix.matrix_gf2e_dense.unpickle_matrix_gf2e_dense_v0`. If m4rie is hidden, or disabled while an extension from a previous build remains importable (the case documented by the new feature), `loads()` therefore reaches the unguarded current function.

This isn't _documented_, that passage is explaining a corner case to avoid test failures. When you're testing the PR by enabling/disabling the feature and running the tests in-place, if you do enabled-then-disabled (in that order), the compiled extension from the "enabled" case will still be sitting in your build directory if you forget to clean it between runs. Having around old versions of files, or deleted files, or files built from another invocation of `meson setup` is not something we would dare try to support. I was just being nice by not including a test that would fail in that case (since it would slow down the review).

File: src/sage/features/m4rie.py (hunk 3 above)
Comment (orlitzky) [reply src/sage/features/m4rie.py:43]: This is a good catch, I didn't realize that `sage.libs.m4ri` was also checking for mr4ie (which depends on m4ri...)

File: src/meson.build (hunk 2 above)
Comment (orlitzky) [reply src/meson.build:182]: > The sage distro is also using m4rie-20250128, and should install a working pc file?

Nevermind, I see what you are saying. If you install m4rie-20150908, the spkg-configure.m4 check will accept it (declining the spkg), but then meson will reject it and disable the feature. You are right but it's telling that you have to go back almost exactly 11 years to find an old enough version to cause this problem :)

File: src/sage/matrix/matrix_space.py (hunk 1 above)
Comment (orlitzky) [reply src/sage/matrix/matrix_space.py:2887]: More details: the test (which mentions this case) is doing a check for runtime detection: if you enabled m4rie, runtime detection should succeed; but if not, runtime detection may still succeed due to build artifacts. To avoid spurious failures, we test only the first scenario, but this does not leak into the code. If you don't enable runtime detection, the runtime detection methods are not normally used, so it doesn't matter if they can "mis-detect" things.

For the unpicking, if you leave `-Ddefer_feature_checks=false` (the default), then the feature check will return whether or not you are allowed to use m4rie features. If you try to restore a pickle from when mr4ie was enabled in a sage where it is disabled, it won't work, but that is intended (why did you disable it?). The converse cannot happen because you cannot make a gf2e pickle if m4rie is disabled.

If OTOH you set `-Ddefer_feature_checks=true`, then whether or not you can unpickle gf2e depends on whether or not you have the gf2e module installed at runtime. If you build it and leave it there and then rebuild with `-Dm4rie=disabled`, then the module is still there, and will be detected at runtime. (This is what you asked for.) Conversely, if the gf2e module is not there at runtime, you won't be able to unpickle gf2e stuff -- but that has not changed.

So AFAICS you can break this, but only by asking for it. (One of the main motivations behind these PRs is to eliminate the "random" runtime behavior changes for normal people, i.e. anyone who does not ask for it.)

File: src/sage/features/m4rie.py (hunk 3 above)
Comment (orlitzky) [reply src/sage/features/m4rie.py:43]: This is done, and nothing seems to have broken.

**Comment by orlitzky** [comment]
I was able to drop the `--force` patch now that linbox has been upgraded, the minimal CI still passes.
